gateleen-routing/src/main/java/org/swisspush/gateleen/routing/DeferCloseHttpClient.java

```diff
@@ -40,65 +42,65 @@ public void request(HttpMethod method, int port, String host, String requestURI,
         countOfRequestsInProgress += 1;
         logger.debug("Pending request count: {}", countOfRequestsInProgress);
-        delegate.request(method, port, host, requestURI).onComplete(handler);
-
-        // TODO: Migrate the original implementation
-        // Delegate to the same method on the delegate. But install our own handler which
-        // allows us to intercept the response.
-//      HttpClientRequest request = delegate.request(method, port, host, requestURI, upstreamRsp -> {
-//            logger.debug("onUpstreamRsp(code={})", upstreamRsp.statusCode());
-//            // 1st we have to pass-through the response so our caller is able to install its handlers.
-//            try {
-//                callHandlerIfExists(responseHandler, upstreamRsp);
-//            } catch (Exception e) {
-//                // Does not make sense to install any handlers. Just make sure we decrement
-//                // our counter then pass-through the exception.
-//                onEndOfRequestResponseCycle();
-//                throw e;
-//            }
-//            // We also need to ensure that our reference counter stays accurate. Badly vertx
-//            // may call BOTH of our handlers. And in this scenario we MUST NOT decrement
-//            // twice. So we additionally track this too.
-//            final AtomicBoolean needToDecrementCounter = new AtomicBoolean(true);
-//            // Then (after client installed its handlers), we now can intercept those by
-//            // replacing them with our own handlers.
-//            // To do this, we 1st backup the original handler (so we can delegate to it later).
-//            Handler<Void> originalEndHandler = getEndHandler(upstreamRsp);
-//            upstreamRsp.endHandler(event -> {
-//                logger.debug("upstreamRsp.endHandler()");
-//                if (needToDecrementCounter.getAndSet(false)) {
-//                    onEndOfRequestResponseCycle();
-//                }
-//                // Call the original handler independent of the above condition to not change
-//                // behaviour of the impl we are decorating.
-//                callHandlerIfExists(originalEndHandler, event);
-//            });
-//            // We also need to intercept exception handler to decrement our counter in case
-//            // of erroneous-end scenario. Basically same idea as above.
-//            Handler<Throwable> originalExceptionHandler = getExceptionHandler(upstreamRsp);
-//            upstreamRsp.exceptionHandler(event -> {
-//                logger.debug("upstreamRsp.exceptionHandler({})", event.toString());
-//                if (needToDecrementCounter.getAndSet(false)) {
-//                    onEndOfRequestResponseCycle();
-//                }
-//                callHandlerIfExists(originalExceptionHandler, event);
-//            });
-//        }
+        delegate.request(method, port, host, requestURI).onComplete(asyncRequestResult -> {
```

Comment:
@hiddenalpha do we need to manually test this in order to know if it still works?

Reply:
If I remind correctly, I wrote the [RuleUpdateTest](https://github.com/swisspush/gateleen/blob/v1.3.0/gateleen-test/src/test/java/org/swisspush/gateleen/routing/RuleUpdateTest.java#L26) for that purpose.

Reply:
The RuleUpdateTest can pass with out this, I mean already passes before enabled this part of code, so start to confuse.

Reply:
The RuleUpdateTest can pass with out this, I mean already passes before enabled this part of code, so start to confuse.

Reply:
I think @ZhengXinCN is right, it passed in v1.3.0 here (the code is commented in this version)

https://api.travis-ci.com/v3/job/561702838/log.txt

[INFO] Running org.swisspush.gateleen.routing.RuleUpdateTest
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 63.751 s - in org.swisspush.gateleen.routing.RuleUpdateTest

maybe the behavior has somehow changed now, but not sure what exactly. I think we should further look into this.